pkg/util/validation/limits.go

```diff
@@ -885,3 +917,93 @@ func EnabledByAnyTenant(tenantIDs []string, f func(string) bool) bool {
 	}
 	return false
 }
+
+// RegisterExtensions registers the extensions type with given name
+// and returns a function to get the extensions from a *Limits instance.
+// The name will be used as YAML/JSON key to decode the extensions.
+func RegisterExtensions[E any](name string) func(*Limits) *E {
+	extensionsType := reflect.TypeOf(new(E))
+	registeredExtensions = append(registeredExtensions, extension{name: name, typ: extensionsType})
+
+	return func(l *Limits) *E {
+		if e, ok := l.extensions[name]; ok {
+			return e.(*E)
+		}
+		return nil
+	}
+}
+
+type extension struct {
+	name string
+	typ  reflect.Type
+}
+
+// registeredExtensions is the list of registered extensions
+var registeredExtensions []extension
```

Comment:
Do we need to protect against concurrent modifications from different goroutines?

Reply:
I would say we don't. We don't want to pay the price of the mutexes if we don't have a use case for that: we don't expect extensions to be added dynamically in runtime right now, just in the package initialization time, so state it as that: https://github.com/grafana/mimir/pull/4397/commits/4ade7fd8e5512a24716574cac1171a656bb402ad

We can add thread safety in the future if we need it.

Reply:
I agree that we only need to support adding new extensions during package initialization time. Just wondering if packages can be initialized from multiple goroutines.

But documentation says:
> Package initialization—variable initialization and the invocation of init functions—happens in a single goroutine, sequentially, one package at a time.

So we're good here.